agent/util-scripts/pbench-tool-meister-start

```diff
@@ -646,186 +552,202 @@ def main(_prog, cli_params):
+    if cli_params.redis_server is None:
+        # FIXME: if only one host is registered, and that host is the same as this
+        # controller, then don't bother starting the Tool Data Sink.
+        logger.debug("4. starting tool data sink")
+        try:
+            pid = os.fork()
+            if pid == 0:
+                # In the child!
+
+                # The main() of the Tool Data Sink module will not return here
+                # since it will daemonize itself and this child pid will be
+                # replaced by a new pid.
+                status = tds_main(
+                    [
+                        PROG.parent / "tool-meister" / "pbench-tool-data-sink",
+                        "localhost",
+                        str(redis_port),
+                        tds_param_key,
+                        "yes",
+                    ]
                 )
-            except Exception:
-                logger.exception("Failed to verify Tool Data Sink logging sink working")
-                return kill_redis_server(redis_pid, ReturnCode.TDSLOGPUBFAILED)
+                sys.exit(status)
             else:
-                if num_present == 0:
-                    time.sleep(0.1)
+                # In the parent!
+
+                # Wait for the child to finish daemonizing itself.
+                retcode = waitpid(pid)
+                if retcode != 0:
+                    logger.error(
+                        "failed to create pbench data sink, daemonized; return code: %d",
+                        retcode,
+                    )
+        except Exception:
+            logger.exception("failed to create pbench data sink, daemonized")
+            return kill_redis_server(redis_pid, ReturnCode.TDSFORKFAILED)
+        else:
+            # Wait for logging channel to be up and ready before we start the
+            # local and remote Tool Meisters.
+            timeout = time.time() + 60
+            num_present = 0
+            while num_present == 0:
+                try:
+                    num_present = redis_server.publish(
+                        f"{cli_tm_channel_prefix}-{tm_channel_suffix_to_logging}",
+                        "pbench-tool-meister-start - verify logging channel up",
+                    )
+                except Exception:
+                    logger.exception(
+                        "Failed to verify Tool Data Sink logging sink working"
+                    )
+                    return kill_redis_server(redis_pid, ReturnCode.TDSLOGPUBFAILED)
+                else:
+                    if num_present == 0:
+                        if time.time() > timeout:
+                            logger.error(
+                                "The Tool Data Sink failed to start within one minute"
+                            )
+                            return kill_redis_server(
+                                redis_pid, ReturnCode.TDSSTARTUPTIMEOUT
+                            )
```

Comment:
If this triggers, then we know that the TDS got far enough to daemonize itself but that it died or got stuck before it could report in.  Shouldn't we at least look and see if it created a PID file that we can use to kill it if it's stuck?  (This seems like a piece of utility code which `pbench-tool-meister-stop` could share.)

Reply:
https://github.com/distributed-system-analysis/pbench/issues/2173